File: src/db/rules_repo.py
"""
Repository for rules table operations.
"""

import sqlite3
from dataclasses import dataclass
from typing import Optional

from src.rules.models import Rule


@dataclass
class RuleRow:
    """Rule record from DB."""

    id: int
    name: str
    file_type: Optional[str]
    regex_pattern: str
    anchor_before: Optional[str]
    anchor_after: Optional[str]
    max_length: Optional[int]
    required: bool
    active: bool


def to_rule(row: RuleRow) -> Rule:
    """Convert RuleRow to Rule model."""
    return Rule(
        id=row.id,
        name=row.name,
        file_type=row.file_type,
        regex_pattern=row.regex_pattern,
        anchor_before=row.anchor_before,
        anchor_after=row.anchor_after,
        max_length=row.max_length,
        required=bool(row.required),
        active=bool(row.active),
    )


def insert(
    conn: sqlite3.Connection,
    name: str,
    regex_pattern: str,
    file_type: Optional[str] = None,
    anchor_before: Optional[str] = None,
    anchor_after: Optional[str] = None,
    max_length: Optional[int] = None,
    required: bool = False,
    active: bool = True,
) -> int:
    """Insert a rule. Returns new id."""
    cursor = conn.execute(
        """
        INSERT INTO rules (name, file_type, regex_pattern, anchor_before, anchor_after, max_length, required, active)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (name, file_type, regex_pattern, anchor_before, anchor_after, max_length, 1 if required else 0, 1 if active else 0),
    )
    conn.commit()
    return cursor.lastrowid


def update(
    conn: sqlite3.Connection,
    rule_id: int,
    *,
    name: Optional[str] = None,
    file_type: Optional[str] = None,
    regex_pattern: Optional[str] = None,
    anchor_before: Optional[str] = None,
    anchor_after: Optional[str] = None,
    max_length: Optional[int] = None,
    required: Optional[bool] = None,
) -> bool:
    """Update rule by id. Returns True if row was updated."""
    updates = []
    params = []
    if name is not None:
        updates.append("name = ?")
        params.append(name)
    if file_type is not None:
        updates.append("file_type = ?")
        params.append(file_type)
    if regex_pattern is not None:
        updates.append("regex_pattern = ?")
        params.append(regex_pattern)
    if anchor_before is not None:
        updates.append("anchor_before = ?")
        params.append(anchor_before)
    if anchor_after is not None:
        updates.append("anchor_after = ?")
        params.append(anchor_after)
    if max_length is not None:
        updates.append("max_length = ?")
        params.append(max_length)
    if required is not None:
        updates.append("required = ?")
        params.append(1 if required else 0)
    if not updates:
        return False
    params.append(rule_id)
    cursor = conn.execute(
        f"UPDATE rules SET {', '.join(updates)} WHERE id = ?",
        params,
    )
    conn.commit()
    return cursor.rowcount > 0


def set_active(conn: sqlite3.Connection, rule_id: int, active: bool) -> bool:
    """Toggle rule active status. Returns True if updated."""
    cursor = conn.execute("UPDATE rules SET active = ? WHERE id = ?", (1 if active else 0, rule_id))
    conn.commit()
    return cursor.rowcount > 0


def delete(conn: sqlite3.Connection, rule_id: int) -> bool:
    """Delete rule. Returns True if deleted."""
    cursor = conn.execute("DELETE FROM rules WHERE id = ?", (rule_id,))
    conn.commit()
    return cursor.rowcount > 0


def get_by_id(conn: sqlite3.Connection, rule_id: int) -> Optional[RuleRow]:
    """Get rule by id."""
    row = conn.execute(
        "SELECT id, name, file_type, regex_pattern, anchor_before, anchor_after, max_length, required, active FROM rules WHERE id = ?",
        (rule_id,),
    ).fetchone()
    if row is None:
        return None
    return RuleRow(
        id=row[0],
        name=row[1],
        file_type=row[2],
        regex_pattern=row[3],
        anchor_before=row[4],
        anchor_after=row[5],
        max_length=row[6],
        required=bool(row[7]),
        active=bool(row[8]),
    )


def list_all(
    conn: sqlite3.Connection,
    active_only: Optional[bool] = None,
    file_type: Optional[str] = None,
) -> list[RuleRow]:
    """List rules, optionally filtered by active and file_type."""
    query = "SELECT id, name, file_type, regex_pattern, anchor_before, anchor_after, max_length, required, active FROM rules WHERE 1=1"
    params = []
    if active_only is not None:
        query += " AND active = ?"
        params.append(1 if active_only else 0)
    if file_type is not None:
        query += " AND (file_type IS NULL OR file_type = ?)"
        params.append(file_type)
    query += " ORDER BY id"
    rows = conn.execute(query, params).fetchall()
    return [
        RuleRow(
            id=r[0],
            name=r[1],
            file_type=r[2],
            regex_pattern=r[3],
            anchor_before=r[4],
            anchor_after=r[5],
            max_length=r[6],
            required=bool(r[7]),
            active=bool(r[8]),
        )
        for r in rows
    ]

File: src/db/rules_repo.cpp
#include "document_ingestion/rules_repo.hpp"
#include <sqlite3.h>
#include <stdexcept>
#include <cstdlib>

namespace document_ingestion {

Rule to_rule(const RuleRow& row) {
  Rule r;
  r.id = row.id;
  r.name = row.name;
  r.file_type = row.file_type;
  r.regex_pattern = row.regex_pattern;
  r.anchor_before = row.anchor_before;
  r.anchor_after = row.anchor_after;
  r.max_length = row.max_length;
  r.required = row.required;
  r.active = row.active;
  return r;
}

int rule_insert(sqlite3* conn, const std::string& name, const std::string& regex_pattern,
                const std::optional<std::string>& file_type,
                const std::optional<std::string>& anchor_before,
                const std::optional<std::string>& anchor_after,
                const std::optional<int>& max_length,
                bool required, bool active) {
  sqlite3_stmt* stmt = nullptr;
  const char* sql = R"(
    INSERT INTO rules (name, file_type, regex_pattern, anchor_before, anchor_after, max_length, required, active)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
  )";
  if (sqlite3_prepare_v2(conn, sql, -1, &stmt, nullptr) != SQLITE_OK)
    throw std::runtime_error(sqlite3_errmsg(conn));

  sqlite3_bind_text(stmt, 1, name.c_str(), -1, SQLITE_TRANSIENT);
  if (file_type)
    sqlite3_bind_text(stmt, 2, file_type->c_str(), -1, SQLITE_TRANSIENT);
  else
    sqlite3_bind_null(stmt, 2);
  sqlite3_bind_text(stmt, 3, regex_pattern.c_str(), -1, SQLITE_TRANSIENT);
  if (anchor_before)
    sqlite3_bind_text(stmt, 4, anchor_before->c_str(), -1, SQLITE_TRANSIENT);
  else
    sqlite3_bind_null(stmt, 4);
  if (anchor_after)
    sqlite3_bind_text(stmt, 5, anchor_after->c_str(), -1, SQLITE_TRANSIENT);
  else
    sqlite3_bind_null(stmt, 5);
  if (max_length)
    sqlite3_bind_int(stmt, 6, *max_length);
  else
    sqlite3_bind_null(stmt, 6);
  sqlite3_bind_int(stmt, 7, required ? 1 : 0);
  sqlite3_bind_int(stmt, 8, active ? 1 : 0);

  if (sqlite3_step(stmt) != SQLITE_DONE) {
    sqlite3_finalize(stmt);
    throw std::runtime_error(sqlite3_errmsg(conn));
  }
  int id = static_cast<int>(sqlite3_last_insert_rowid(conn));
  sqlite3_finalize(stmt);
  sqlite3_exec(conn, "COMMIT", nullptr, nullptr, nullptr);
  return id;
}

std::vector<RuleRow> rules_list_all(sqlite3* conn, const std::optional<bool>& active_only,
                                    const std::optional<std::string>& file_type) {
  std::string sql = "SELECT id, name, file_type, regex_pattern, anchor_before, anchor_after, max_length, required, active FROM rules WHERE 1=1";
  if (active_only) sql += " AND active = ?";
  if (file_type) sql += " AND (file_type IS NULL OR file_type = ?)";
  sql += " ORDER BY id";

  sqlite3_stmt* stmt = nullptr;
  if (sqlite3_prepare_v2(conn, sql.c_str(), -1, &stmt, nullptr) != SQLITE_OK)
    return {};

  int idx = 1;
  if (active_only) sqlite3_bind_int(stmt, idx++, *active_only ? 1 : 0);
  if (file_type) sqlite3_bind_text(stmt, idx++, file_type->c_str(), -1, SQLITE_TRANSIENT);

  std::vector<RuleRow> result;
  while (sqlite3_step(stmt) == SQLITE_ROW) {
    RuleRow row;
    row.id = sqlite3_column_int(stmt, 0);
    row.name = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
    if (sqlite3_column_type(stmt, 2) != SQLITE_NULL)
      row.file_type = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2));
    row.regex_pattern = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 3));
    if (sqlite3_column_type(stmt, 4) != SQLITE_NULL)
      row.anchor_before = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 4));
    if (sqlite3_column_type(stmt, 5) != SQLITE_NULL)
      row.anchor_after = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 5));
    if (sqlite3_column_type(stmt, 6) != SQLITE_NULL)
      row.max_length = sqlite3_column_int(stmt, 6);
    row.required = sqlite3_column_int(stmt, 7) != 0;
    row.active = sqlite3_column_int(stmt, 8) != 0;
    result.push_back(row);
  }
  sqlite3_finalize(stmt);
  return result;
}

}  // namespace document_ingestion
